### PR TITLE
add zero grad tracking

### PR DESCRIPTION
**Pull request by faresobeid**
Screenshot 2026-03-06 at 05 42 24

---

> [!NOTE]
> **Low Risk**
> Adds a read-only training metric and logging; main risk is minor overhead or incorrect aggregation under distributed sharding/DP replication.
> 
> **Overview**
> Adds a new *zero-gradient ratio* observability metric that counts (and all-reduces) the fraction of parameter elements with zero/absent gradients, including handling for `DTensor` sharded parameters and parameters with `grad is None`.
> 
> Both RL (`trainer/rl/train.py`) and SFT (`trainer/sft/train.py`) training loops now compute `optim/zero_grad_ratio` each step and log it to the monitor, and the Prometheus `MetricsServer` exposes it via a new `trainer_zero_grad_ratio` gauge included in `MetricsServer.update()`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 0b69544322e6b79ea03cf8682f9e14899d8a80cc. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/rl/train.py [hunk 1]
```diff
@@ -480,6 +482,13 @@ def load_run_checkpoint(_optimizer, idx: int) -> None:
 
         # Synchronize the tensor metrics across all steps and ranks
         tensor_stats = tensors.compute_stats()
+        dist.all_reduce(num_zero_grad, op=dist.ReduceOp.SUM)
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/rl/train.py]: hm, we have the Metrics class for this on the trainer no?

File: src/prime_rl/trainer/rl/train.py [hunk 2]
```diff
@@ -497,7 +506,7 @@ def load_run_checkpoint(_optimizer, idx: int) -> None:
         step_message = f"Step {progress.step} | Time: {step_time:.2f}s | Loss: {tensor_stats['loss/mean']:.4f} | Entropy: {tensor_stats['entropy/mean']:.4f}"
         if "mismatch_kl/mean" in tensor_stats:
             step_message += f" | Mismatch KL: {tensor_stats['mismatch_kl/mean']:.4f}"
-        step_message += f" | Grad. Norm: {grad_norm:.4f} | LR: {current_lr:.2e} | Throughput: {throughput:.0f} tokens/s | MFU: {mfu:.1f}% | Peak Mem.: {peak_memory:.1f} GiB"
+        step_message += f" | Grad. Norm: {grad_norm:.4f} | Num Zero Grad: {num_zero_grad.item():,} | LR: {current_lr:.2e} | Throughput: {throughput:.0f} tokens/s | MFU: {mfu:.1f}% | Peak Mem.: {peak_memory:.1f} GiB"
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/rl/train.py]: not sure its important enough to be in step message?

File: src/prime_rl/trainer/rl/train.py [hunk 3]
```diff
@@ -516,6 +525,8 @@ def load_run_checkpoint(_optimizer, idx: int) -> None:
         optim_metrics = {
             "optim/lr": current_lr,
             "optim/grad_norm": grad_norm.item(),
+            "optim/num_zero_grad": num_zero_grad.item(),
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/rl/train.py]: is this absolute num params? could remove this imo

**Review by mikasenghaas — APPROVED** [review]
nice, lot cleaner. lgtm!

**Review by samsja — APPROVED** [review]
